src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -1136,6 +1137,53 @@ public SolutionCompilationState WithOptions(SolutionOptionSet options)
             this.SolutionState.WithOptions(options));
     }
 
+    public SolutionCompilationState WithSourceGeneratorVersions(
+        FrozenDictionary<ProjectId, SourceGeneratorExecutionVersion> projectIdToSourceGeneratorExecutionVersion, CancellationToken cancellationToken)
+    {
+        var newIdToProjectStateMapBuilder = this.SolutionState.ProjectStates.ToBuilder();
+        var newIdToTrackerMapBuilder = _projectIdToTrackerMap.ToBuilder();
+        var changed = false;
+
+        foreach (var (projectId, sourceGeneratorExecutionVersion) in projectIdToSourceGeneratorExecutionVersion)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var projectState = newIdToProjectStateMapBuilder[projectId];
+
+            // Nothing to do if already at this version.
+            if (projectState.Attributes.SourceGeneratorExecutionVersion == sourceGeneratorExecutionVersion)
+                continue;
+
+            changed = true;
+            newIdToProjectStateMapBuilder[projectId] = projectState.WithSourceGeneratorExecutionVersion(sourceGeneratorExecutionVersion);
+
+            // If we do already have a compilation tracker for this project, then let the tracker know that the source
+            // generator version has changed. We do this by telling it that it should now create SG docs and skeleton
+            // references if they're out of date.
+            if (_projectIdToTrackerMap.TryGetValue(projectId, out var existingTracker))
+            {
+                // if the major version has changed then we also want to drop the generator driver so that we're rerun
+                // generators from scratch.
+                var forceRegeneration = projectState.Attributes.SourceGeneratorExecutionVersion.MajorVersion != sourceGeneratorExecutionVersion.MajorVersion;
+                var newTracker = existingTracker.WithCreationPolicy(create: true, forceRegeneration, cancellationToken);
```

Comment:
changing the SG version 'unfreezes' the compilation tracker, so it can regen sg docs.